pkg/skaffold/build/scheduler.go

```diff
@@ -35,45 +35,58 @@ type ArtifactBuilder func(ctx context.Context, out io.Writer, artifact *latest.A
 
 // For testing
 var (
-	buffSize      = bufferedLinesPerArtifact
-	runInSequence = InSequence
+	buffSize = bufferedLinesPerArtifact
 )
 
-// InParallel builds a list of artifacts in parallel but prints the logs in sequential order.
-func InParallel(ctx context.Context, out io.Writer, tags tag.ImageTags, artifacts []*latest.Artifact, buildArtifact ArtifactBuilder, concurrency int) ([]Artifact, error) {
-	if len(artifacts) == 0 {
-		return nil, nil
-	}
+type artifactWithDeps struct {
+	artifact              *latest.Artifact
+	artifactChan          chan interface{}
+	requiredArtifactChans []chan interface{}
+}
 
-	if len(artifacts) == 1 || concurrency == 1 {
-		return runInSequence(ctx, out, tags, artifacts, buildArtifact)
+func (a artifactWithDeps) markComplete() {
+	// closing channel notifies all listeners waiting for this build to complete
+	close(a.artifactChan)
+}
+func (a artifactWithDeps) waitForDependencies(ctx context.Context) {
+	for _, dep := range a.requiredArtifactChans {
+		// wait for dependency to complete build
+		select {
+		case <-ctx.Done():
+		case <-dep:
+		}
 	}
+}
 
+// InOrder builds a list of artifacts in dependency order within the specified max concurrency
+func InOrder(ctx context.Context, out io.Writer, tags tag.ImageTags, artifacts []*latest.Artifact, buildArtifact ArtifactBuilder, concurrency int) ([]Artifact, error) {
+	awdSlice := artifactWithDepsFor(artifacts)
 	var wg sync.WaitGroup
 	defer wg.Wait()
 
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
 
 	results := new(sync.Map)
-	outputs := make([]chan string, len(artifacts))
+	outputs := make([]chan string, len(awdSlice))
 
 	if concurrency == 0 {
-		concurrency = len(artifacts)
+		concurrency = len(awdSlice)
 	}
 	sem := make(chan bool, concurrency)
 
-	// Run builds in //
 	wg.Add(len(artifacts))
-	for i := range artifacts {
+	for i := range awdSlice {
 		outputs[i] = make(chan string, buffSize)
 		r, w := io.Pipe()
 
 		// Run build and write output/logs to piped writer and store build result in
 		// sync.Map
 		go func(i int) {
+			awdSlice[i].waitForDependencies(ctx)
 			sem <- true
-			runBuild(ctx, w, tags, artifacts[i], results, buildArtifact)
+			runBuild(ctx, w, tags, awdSlice[i].artifact, results, buildArtifact)
```

Comment:
So what happens if a dependency build failed?

Reply:
So this PR is just parity with existing behavior. So even if a build dependency fails it doesn't fails the downstream builds since the artifacts aren't being cross-referenced anywhere yet. However in my local branch I'm maintaining additional state in this controller which causes subsequent builds to fail since a required artifact build was not found.

Reply:
Added a test case for failure. The current behavior is to just report on the first failed build and I haven't modified it.